### ISSUE TITLE
Did something change lately?

### ISSUE THREAD

**Issue by gustavogx** [dandistine/olcPGEDearImGui#3]
Hi, I spent a few hours now, trying to compile the example, without success. It should be quite simple, right? I tried the includes you suggested and also others which I know should work (I also have a working example of imgui running on a glfw window, and it compiles fine). Just as a note, I also have olc::PGE examples working. But I can't compile yours. Does it also need the Stb library? 

Anyway, I know you don't want to provide a compiler-system example (like make or cmake) and I understand that. But it would be really useful to provide a compiling example with the right command line in g++ or clang, just like Javix did with the olc::PGE itself. What do you think?

Thank you.

**Comment by dandistine** [comment]
As you haven't provided any information as to what your issue actually is, I can only make guesses.  I would suppose that you did not include the required files from Dear ImGui in your build.  When using the default renderer in PGE there are no additional dependencies.  When using OpenGL 3.3 by specifying PGE_GFX_OPENGL33 you need to have GLEW available and add it to your linker options (something like -lglew32 probably).

**Comment by gustavogx** [comment]
I apologize. It was super late for me and I was lazy explaining myself. Anyway, I fixed. Take a look:

Below you'll find my example.cpp, which is the one you suggested in the comments section of you .h file, but with a couple of changes: 1) I had to add a #define IMGUI_DEFINE_MATH_OPERATORS before including so the implementations of the several operator() could be generated, and 2) had to also include #include "imgui_tables.cpp" to be compiled. Keep in mind I am using the latest ImGui v. 1.86 and PGE v. 2.16. 

That was it. I compiled it with 

g++ example.cpp -o example -lX11 -lGL -lpthread -lpng -lstdc++fs -std=c++17 -Iimgui -Iimgui/backends

and it ran smoothly:

![Screenshot from 2021-11-09 10-11-51](https://user-images.githubusercontent.com/13077868/140930318-f4d72e21-d763-4c70-8c24-144879058c4e.png)

If you uncomment #define PGE_GFX_OPENGL33, it should work as well (no need to change the compiler line).

Thank you for your attention and for this great work. Feel free to add this corrections to your example or to use mine as a complete file. 


// ===========================================
//
// example.cpp
// To be used with https://github.com/dandistine/olcPGEDearImGui
//
// compatible with  
//       https://github.com/ocornut/imgui v. 1.86
//       https://github.com/OneLoneCoder/olcPixelGameEngine v. 2.16
//
//
// Place DearImGui in a folder (imgui) alongside this example.
// Your folder should look like this:
//
// imgui
// example.cpp
// imgui_impl_pge.h
// olcPixelGameEngine.h
//
// Under linux, you may compile it like this:
// g++ example.cpp -o example -lX11 -lGL -lpthread -lpng -lstdc++fs -std=c++17 -Iimgui -Iimgui/backends
//
// ===========================================

#include "imgui.h"
#define IMGUI_DEFINE_MATH_OPERATORS
#include "imgui_internal.h"
#include "imstb_rectpack.h"
#include "imstb_textedit.h"
#include "imstb_truetype.h"

#include "imgui.cpp"
#include "imgui_demo.cpp"
#include "imgui_draw.cpp"
#include "imgui_widgets.cpp"
#include "imgui_tables.cpp"
#include "imgui_impl_opengl2.cpp"
#include "imgui_impl_opengl2.h"

#define OLC_PGEX_DEAR_IMGUI_IMPLEMENTATION
//#define PGE_GFX_OPENGL33
#include "imgui_impl_pge.h"

#define OLC_PGE_APPLICATION
#include "olcPixelGameEngine.h"

class Example : public olc::PixelGameEngine
{
	olc::imgui::PGE_ImGUI pge_imgui;
	int m_GameLayer;
public:
	//PGE_ImGui can automatically call the SetLayerCustomRenderFunction by passing
	//true into the constructor.  false is the default value.
	Example() : pge_imgui(false)
	{
		sAppName = "Test Application";
	}
public:
	bool OnUserCreate() override
	{
		//Create a new Layer which will be used for the game
		m_GameLayer = CreateLayer();
		//The layer is not enabled by default,  so we need to enable it
		EnableLayer(m_GameLayer, true);
		//Set a custom render function on layer 0.  Since DrawUI is a member of
		//our class, we need to use std::bind
		//If the pge_imgui was constructed with _register_handler = true, this line is not needed
		SetLayerCustomRenderFunction(0, std::bind(&Example::DrawUI, this));
		return true;
	}
	bool OnUserUpdate(float fElapsedTime) override
	{
		//Change the Draw Target to not be Layer 0
		SetDrawTarget((uint8_t)m_GameLayer);
		//Game Drawing code here
		//Create and react to your UI here, it will be drawn during the layer draw function
		ImGui::ShowDemoWindow();
		return true;
	}
	void DrawUI(void) {
		//This finishes the Dear ImGui and renders it to the screen
		pge_imgui.ImGui_ImplPGE_Render();
	}
};

int main() {
	Example demo;
	if (demo.Construct(640, 360, 2, 2))
		demo.Start();
	return 0;
}

// ===========================================

**Comment by dandistine** [comment]
I can see now why your build was not originally working.

You're including way more files in your file than actually necessary.  All of the cpp files that you #include should actually be compiled separately as part of your build process.  You can do this by adding them to the command:

`g++ example.cpp imgui.cpp imgui_demo.cpp imgui_draw.cpp` and so on with each cpp file making an appearance.


The `#define IMGUI_DEFINE_MATH_OPERATORS` is required in your specific file because you are using #include on the cpp files after you `#include imgui_internal.h`.  If you take a look at the imgui cpp files, you'll see that they already internally include imgui_internal.h (and the other imgui headers) after defining that macro if required.

The addition of the imgui_tables.cpp file is because that file did not exist when I last updated the extension.  I don't really track the development of Dear ImGui closely, I have a copy (1.76) which I use.

You can remove all the lines between the end of your top block comment and your #define OLC_PGEX_DEAR_IMGUI_IMPLEMENTATION.  Your build command after removing all of that would look something like this:

`g++ example.cpp imgui/imgui.cpp imgui/imgui_demo.cpp imgui/imgui_draw.cpp imgui/imgui_widgets.cpp imgui/imgui_tables.cpp imgui/imgui_impl_opengl2.cpp -o example -lX11 -lGL -lpthread -lpng -lstdc++fs -std=c++17 -Iimgui -Iimgui/backends`

**Comment by gustavogx** [comment]
Yeah, thank you, you are absolutely right (specially about the cpp). I did that mostly so I could simplify the compilation, but it actually got worse, right? I will try again.

BTW, I forked your repository and added this example (which I will now fix it). Right now it is public, but I can set it to private if it bothers you. OR I can just have the example without the copy of your .h file. You name it.
Again, thank you for your contribution.
